### DISCUSSION TITLE
Change of variables for functions with invariant integral

### DISCUSSION THREAD

**Discussion by cmichelenstrofer** [sandialabs/WecOptTool#106] [Theory]
Some functions are defined by properties of their integral, and their integral over a differential area must be invariant over a change of variables. That is, a function <img src="https://render.githubusercontent.com/render/math?math=f"> is defined such that <img src="https://render.githubusercontent.com/render/math?math=\int_a^b f(x) dx = C">, and the integral over the same area (a-b) needs to be the same constant C even after change of variables. These functions include:

- The **_elevation variance spectrum_**, <img src="https://render.githubusercontent.com/render/math?math=S(f)">, where <img src="https://render.githubusercontent.com/render/math?math=\int_a^b S(f) df">   is the total variance in the frequency range <img src="https://render.githubusercontent.com/render/math?math=f\in[a,b]">.
- The **_energy spectrum_**, <img src="https://render.githubusercontent.com/render/math?math=\rho g S(f)"> for gravity waves, where the integral gives the total energy in a frequency range.
- The **_spread function_**, <img src="https://render.githubusercontent.com/render/math?math=D(f, \theta)">, where the integral <img src="https://render.githubusercontent.com/render/math?math=\int_a^b D(f, \theta) d\theta = C"> gives the percentage of the total energy in freuquency f that is contained in the range of directions <img src="https://render.githubusercontent.com/render/math?math=\theta\in[a,b]">. 
- [Probability density functions](https://en.wikipedia.org/wiki/Probability_density_function#Function_of_random_variables_and_change_of_variables_in_the_probability_density_function) where the integral gives the probability for a range of the random variable (not used in this code but analogous problem and the Wikipedia article has a good explanation.) 

If we have a function f such that its integral over a range is a constant C:
<img src="https://render.githubusercontent.com/render/math?math=\int_{x_0}^{x_f}f\left(x\right)dx=C"> 
and we want to change the variables from x to y, where x and y are related by the invertible function g as 
<img src="https://render.githubusercontent.com/render/math?math=y=g(x)"> 
and 
<img src="https://render.githubusercontent.com/render/math?math=dy=g^\prime(x)dx">  
The change of variables is
<img src="https://render.githubusercontent.com/render/math?math=\int_{x_0}^{x_f}f\left(x\right)dx=C=\int_{g(x_0)}^{g(x_f)}f\left(g^{-1}(y)\right)\frac{1}{g^\prime(y)}dy"> 

**Example: _Spread function radians to degrees_.**
If we have the spread function as a function of angle in radians <img src="https://render.githubusercontent.com/render/math?math=D_r(f, \theta_r)"> and want it in terms of angle in degrees <img src="https://render.githubusercontent.com/render/math?math=D_d(f, \theta_d)"> the function g is <img src="https://render.githubusercontent.com/render/math?math=\theta_d=g(\theta_r)=\frac{180}{\pi}\theta_r"> and <img src="https://render.githubusercontent.com/render/math?math=g^\prime(\theta_r)=\frac{180}{\pi}">. This gives

<img src="https://render.githubusercontent.com/render/math?math=D_d(f, \theta_d) = \frac{\pi}{180}D_r(f, \frac{\pi}{180}\theta_d)">

The opposite situation is 
<img src="https://render.githubusercontent.com/render/math?math=D_r(f, \theta_r) = \frac{180}{\pi}D_d(f, \frac{180}{\pi}\theta_r)">

**Example: _Elevation variance spectrum radians to Hz_.**
By similar arguments the result is

<img src="https://render.githubusercontent.com/render/math?math=S_f(f) = 2\pi S_\omega(2\pi f)"> 

The opposite situation is
<img src="https://render.githubusercontent.com/render/math?math=S_\omega(\omega) = \frac{1}{2\pi}S_f(\frac{1}{2\pi}\omega)"> 

These also apply to the energy spectrum, simply carry the factor <img src="https://render.githubusercontent.com/render/math?math=\rho g"> (for gravity waves).

**Example: _Elevation variance spectrum frequency to period_.**
In this case the function g is not a linear function, but applying the same methodology for period T in seconds and frequency f in Hz we get

<img src="https://render.githubusercontent.com/render/math?math=S_T(T) = \frac{-1}{T^2}S_f(\frac{1}{T})">

The opposite situation is
<img src="https://render.githubusercontent.com/render/math?math=S_f(f) = \frac{-1}{f^2}S_T(\frac{1}{f})">

The negative sign is due to the order of the integration bounds being reversed. We could make a definition decision here so that <img src="https://render.githubusercontent.com/render/math?math=S_T(T)"> is positive. 

**TLDR;**
For functions whose integral must be invariant over change of variables it is not sufficient to convert the input of the function. For instance when converting a plot of (w, S(w)) to (f, S(f)), it is not only the x-axis that needs to be scaled, the y-axis will have have a scaling too. Similar for changing the spread function between inputs in radians and degrees. The case for converting the spectrum from frequency input to period input is a bit more complicated, see above.

**Comment by cmichelenstrofer** [comment]
The required scaling is pretty obvious for the cases where it is only a change of units.  E.g. for S(f) with f in rad/s to Hz if you consider the units of both the frequency and the spectrum and the fact that the units of frequency times the unit of the spectrum need to give you units of elevation variance (m^2) in the case of the elevation variance spectrum. Similar to the spread function where the product of units needs to be non-dimensional. The conversion between f and T is not as trivial.

**Comment by cmichelenstrofer** [comment]
[change_of_variables.pdf](https://github.com/SNL-WaterPower/WecOptTool/files/8557191/change_of_variables.pdf)
